create_functions.sql

```diff
@@ -417,56 +510,170 @@ BEGIN
+CREATE OR REPLACE PROCEDURE update_older_account(
+    max_slot BIGINT
+)
 
-DECLARE
-    branch_slots BIGINT[];
+AS $update_older_account$
 
 BEGIN
-    WITH RECURSIVE parents(slot, parent, status, depth) AS (
+    -- add recent states of all accounts from account_audit
+    -- before slot max_slot into older_account table
+    INSERT INTO public.older_account
+    SELECT
+        acc1.pubkey,
+        acc1.owner,
+        acc1.lamports,
+        acc1.slot,
+        acc1.executable,
+        acc1.rent_epoch,
+        acc1.data,
+        acc1.write_version,
+        acc1.updated_on,
+        acc1.txn_signature
+    FROM public.account_audit AS acc1
+    INNER JOIN (
         SELECT
-            first.slot,
-            first.parent,
-            first.status,
-            1
-        FROM slot AS first
-        WHERE first.slot = start_slot
-        UNION
-            SELECT
-                next.slot,
-                next.parent,
-                next.status,
-                depth + 1
-            FROM slot AS next
-            INNER JOIN parents p ON p.parent = next.slot
-            WHERE depth < num_hashes
-    )
-    SELECT array_agg(prnts.slot)
-    INTO branch_slots
-    FROM parents AS prnts;
+            acc2.pubkey AS pubkey,
+            MAX(acc2.slot) AS slot,
+            MAX(acc2.write_version) AS write_version
+        FROM public.account_audit AS acc2
+        WHERE acc2.slot < max_slot
+        GROUP BY acc2.pubkey
+    ) latest_versions
+    ON
+        latest_versions.pubkey = acc1.pubkey
+        AND latest_versions.slot = acc1.slot
+        AND latest_versions.write_version = acc1.write_version
+    WHERE
+        acc1.slot < max_slot
+    ON CONFLICT (pubkey) DO UPDATE SET 
+		slot=excluded.slot, 
+		owner=excluded.owner, 
+		lamports=excluded.lamports, 
+		executable=excluded.executable, 
+		rent_epoch=excluded.rent_epoch,
+        data=excluded.data, 
+		write_version=excluded.write_version, 
+		updated_on=excluded.updated_on, 
+		txn_signature=excluded.txn_signature;
+END;
+$update_older_account$ LANGUAGE plpgsql;
+
+-----------------------------------------------------------------------------------------------------------------------
+CREATE OR REPLACE FUNCTION get_recent_update_slot(
+    in_pubkey BYTEA,
+    max_slot BIGINT
+)
 
+RETURNS TABLE (
+	slot BIGINT
+)
+
+AS $get_recent_update_slot$
+
+BEGIN
     RETURN QUERY
-        SELECT b.slot, b.blockhash
-        FROM block AS b
-        WHERE b.slot = ANY(branch_slots);
+        WITH results AS (
+            SELECT acc.slot, acc.write_version 
+            FROM public.account_audit AS acc
+            WHERE 
+                acc.pubkey = in_pubkey 
+                AND acc.slot <= max_slot 
+            UNION
+            SELECT old_acc.slot, old_acc.write_version
+            FROM public.older_account AS old_acc
+            WHERE
+                old_acc.pubkey = in_pubkey
+        )
+        SELECT res.slot
+        FROM results AS res
```

Comment:
"res.write_version DESC" is unnecessary. Write_version doesn't affect the result set.

Reply:
[discussed]